pkg/controller/shared_state.go

```diff
@@ -1345,13 +1373,152 @@ func (c *SharedState) createReplicas(ctx context.Context, remainingReplicas int,
 		//no failed replica attachments, but there are still more replicas to reach MaxShares
 		request := ReplicaRequest{VolumeName: volumeName, Priority: remainingReplicas}
 		c.priorityReplicaRequestsQueue.Push(ctx, &request)
-		for _, pod := range pods {
-			c.eventRecorder.Eventf(pod.DeepCopyObject(), v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, "Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		message := fmt.Sprintf("Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		podCopies := make([]runtime.Object, len(pods))
+		for i, pod := range pods {
+			podCopies[i] = pod.DeepCopyObject()
 		}
+		timestamp := time.Now()
+		for _, podCopy := range podCopies {
+			c.eventRecorder.Eventf(podCopy, v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, message)
+		}
+		message += timestamp.UTC().Format(" [0102 15:04:05") // append original timestamp
+		c.persistAttachmentFailure(volumeName, message, podCopies, timestamp)
 	}
 	return nil
 }
 
+func (c *SharedState) persistAttachmentFailure(volumeName string, message string, podCopies []runtime.Object, timestamp time.Time) {
+	c._eventRefresherInitOnce.Do(c._eventRefresherInitFunc)
+	c.eventsToPersistQueue <- ReplicaAttachmentFailureInfo{volumeName, message, podCopies, timestamp}
+}
+
+// Only requests an event un-persist if the event refresher is active
+func (c *SharedState) unpersistAttachmentFailure(volumeName string) {
+	trySend(c.eventsToUnpersistQueue, volumeName) // also mitigates unlikely race condition of sending to closed channel
+}
+
+func (c *SharedState) _eventRefresherInitFunc() {
+	c.eventsToPersistQueue = make(chan ReplicaAttachmentFailureInfo, maxPersistEventConcurrency)
+	c.eventsToUnpersistQueue = make(chan string, maxUnpersistEventConcurrency)
+	go c._eventRefresherRoutine()
+}
+
+func (c *SharedState) _eventRefresherRoutine() {
+	var eventTTL = time.Duration(c.config.EventTTLInSec) * time.Second
+	eventMap := map[string]*eventRefresherNode{}
+	newFailureInfo := <-c.eventsToPersistQueue
+	klog.V(2).Infof("Starting the event refresher")
+	var events eventRefresherList
+	var alarm *time.Timer
+	var lastTime time.Time
+	var delay time.Duration            // how long the alarm was last set to wait for
+	expLatency := eventOverlapVariance // we request the alarm to wake us up earlier by this amount, to negate latency due to other operations and timer imprecision
+	for {
+		// add the new failure
+		newEvent := &eventRefresherNode{
+			message:   newFailureInfo.message,
+			objects:   newFailureInfo.pods,
+			timestamp: newFailureInfo.timestamp.Add(eventTTL - eventOverlapDuration),
+		}
+
+		if events.isEmpty() {
+			lastTime = time.Now()
+			delay = newEvent.timestamp.Sub(lastTime) - expLatency
+			alarm = time.NewTimer(delay)
```

Comment:
On first activation `delay` is `duration.Zero`, so the alarm will fire immediately and add a duplicate event?

Reply:
`{event}.timestamp` is when the event should next be refreshed, initialized to the original timestamp plus the ttl, etc